### PR TITLE
Extend Scala language

### PR DESCRIPTION
**Pull request by 2d6**
This pull request intends to extend the scala language definitions introduced in #735. If applied, it will

- add further types
- add modifiers
- add keywords
- remove added `enter` after `else` keyword to enable `if(foo) bar else baz` syntax

### CODE REVIEW

File: lang/scala/scala.talon
Comment (splondike) [inline lang/scala/scala.talon:27]: I think I'd keep the state prefix required for these ones, we tend to stay away from single word commands in knauj public to reduce the chance of collisions with future commands and also because they tend to recognise less well.

**Comment by splondike** [comment]
@SCdF Happy to lose that enter press after the else statement?

File: lang/scala/scala.py
Comment (SCdF) [inline lang/scala/scala.py:65]: so `def` is how you define a function and so is already covered by "funk <blah>", as is I think `package` and possibly others below. @splondike not sure about default repo's philosophy regarding duplicate ways of doing things, but thought I'd flag if you weren't aware.

**Comment by SCdF** [comment]
@splondike yep happy with single line else, it's definitely more correct! Made one comment about duplicate functionality as well

File: lang/scala/scala.py
Comment (2d6) [reply lang/scala/scala.py:65]: This is mostly true, but at least in our codebase, we also sometimes use `def` for expressions without argument brackets:
```scala
def foo = "hello"
// or
override def bar = "baz"
```
afaik, `funky` inserts `()` automatically. So I'm not sure this counts as duplicate functionality. What do you think?

File: lang/scala/scala.talon
Comment (2d6) [reply lang/scala/scala.talon]: Good point, thank you! I amended this in the commit.

**Comment by 2d6** [comment]
Now that I think of it @splondike : in one of my pull requests, you mentioned that for code we should use `insert` instead of `auto_insert` (https://github.com/knausj85/knausj_talon/pull/752#r795277190). If this also applies here, I would be happy to amend that here, as well. There are several operator methods which use `auto_insert` in `scala.py`. Just let me know :)

**Comment by 2d6** [comment]
I think it would also make sense to extend the language mappings with the `.sbt` extension, because the scala build tool also uses scala syntax for the `build.sbt`, `plugins.sbt` etc. files.

File: lang/scala/scala.py
Comment (SCdF) [reply lang/scala/scala.py:64]: Yep that's fair. Scala is nothing if not flexible in how you use it

**Comment by SCdF** [comment]
> I think it would also make sense to extend the language mappings with the `.sbt` extension, because the scala build tool also uses scala syntax for the `build.sbt`, `plugins.sbt` etc. files.

This is a good example of scala's flexibility making building this kind of thing really hard. The sbt file is a dsl based on scala, but not really idiomatic scala, right? So a build file can look like:

```sbt
ThisBuild / organization := "com.example"
ThisBuild / scalaVersion := "2.12.15"
ThisBuild / version      := "0.1.0-SNAPSHOT"

lazy val root = (project in file("."))
  .settings(
    name := "hello"
  )
```
Which is sort of scala, but also you're not "dividing" ThisBuild by organisation for example, so if you wanted an idiomatic sbt mapping you would want to call that something else, and that new mapping would only apply to sbt files. Similarly with  `:=`, in  sbt you could map this to "op assign" wherein scala that would just be  `=`.

If you think the value of having sbt use scala definitions is additive to your experience though, even if it's a little wrong, then that's probably OK?

There are tons of DSLs in scalaland unfortunately, but at least this one is mostly scala and most people are guaranteed to use it.

_(FWIW I also did the cursorless mapping for scala, and I did not even look into sbt files or how they get parsed out, so it's likely there is work required for support there, if you use that tool)_

**Comment by splondike** [comment]
Yeah, you can replace the auto_inserts with insert to be a little more correct IMO, good memory!

For the SBT stuff it sounds like maybe that should be a different language (and probably a different PR)? You could do something like has been done with typescript/javascript perhaps. Activate the user.scala tag from the sbt language and then override some select things (e.g. op assign -> ':=').

**Comment by 2d6** [comment]
> Yeah, you can replace the auto_inserts with insert to be a little more correct IMO, good memory!
Thank you, I amended this to the commit.

> For the SBT stuff it sounds like maybe that should be a different language (and probably a different PR)? You could do something like has been done with typescript/javascript perhaps. Activate the user.scala tag from the sbt language and then override some select things (e.g. op assign -> ':=').
This actually sounds like a great idea. Thanks for this suggestion :)

Also thank both of you for your valuable suggestions and feedback. Unless you find anything else to improve in this pull request, I currently have nothing else to add to it.

File: lang/scala/scala.py
Comment (rntz) [inline lang/scala/scala.py:14]: "short" seems to have been lost, is this intentional?

File: lang/scala/scala.py
Comment (rntz) [inline lang/scala/scala.py:40]:
```suggestion
    "seek": "Seq",
```

File: lang/scala/scala.talon
Comment (rntz) [inline lang/scala/scala.talon:32]: It's weird that modifiers get space added to them by the command while keywords include the space in them. It would be clearer if it was consistent.

**Review by rntz — APPROVED** [review]
Looks mostly good to me, however, looks like "short" has been lost from the scala_common_types list. Is that intentional?

File: lang/scala/scala.py
Comment (2d6) [reply lang/scala/scala.py:14]: D'oh, good catch! I amended this.

**Comment by rntz** [comment]
I'm just going to merge this, if you want to address any further concerns such as the short type or the comments I left then just make another PR.

File: lang/scala/scala.py
Comment (2d6) [reply lang/scala/scala.py]: Good suggestion. I also modified the command for `Deque` from `deque` to `deck`.

File: lang/scala/scala.talon
Comment (2d6) [reply lang/scala/scala.talon]: You are absolutely right. I have now included this in the commit.

**Comment by rntz** [comment]
oh wow, looks like you got to my concerns before I merged : )

**Comment by 2d6** [comment]
Phew, just in time :D Sorry for involuntarily sneaking in the changes, but I'm happy to see the pull request merged!

**Pull request merged**
